tests/untl_structure_test.py

```diff
@@ -0,0 +1,490 @@
+    # Check that the first child's text is the name.
+    assert subelement[0].text == name.content
+
+
+def test_add_missing_children():
+    """Test adding expected children that don't exist for a form."""
+    required = ['title', 'format', 'publisher']
+    parent = us.UNTLElement()
+    parent.contained_children = required
+    title_child = us.Title(content='A title',
+                           qualifier='officialtitle')
+    children = [title_child]
+    parent.children = children
+    padded_children = us.add_missing_children(required, children)
+    # Now there are Format and Publisher elements after Title.
+    assert len(padded_children) == 3
+    assert isinstance(padded_children[1], us.Format)
+    assert isinstance(padded_children[2], us.Publisher)
+
+
+def test_UNTLElement_init():
+    """Check initialized UNTLElement."""
+    element = us.UNTLElement(content='test_content',
+                             qualifier='test_qualifier')
+    assert element.tag is None
+    assert element.contained_children == []
+    assert element.allows_content
+    assert element.allows_qualifier
+    assert element.qualifier == 'test_qualifier'
+    assert element.children == []
+    assert element.content == 'test_content'
+
+
+def test_UNTLElement_set_qualifier():
+    """Test setting and stripping of qualifier value."""
+    element = us.UNTLElement()
+    element.set_qualifier('  test_qualifier ')
+    assert element.qualifier == 'test_qualifier'
+
+
+def test_UNTLElement_set_qualifier_exception():
+    """Test a qualifier must be allowed to set one."""
+    element = us.UNTLElement()
+    element.allows_qualifier = False
+    with pytest.raises(us.UNTLStructureException):
+        element.set_qualifier('test_qualifier')
+
+
+def test_UNTLElement_add_child():
+    """Test children in contained_children are allowed for adding."""
+    child_element = us.UNTLElement()
+    child_tag = 'child1'
+    child_element.tag = child_tag
+    element = us.UNTLElement()
+    element.contained_children.append(child_tag)
+    element.add_child(child_element)
+    assert child_element in element.children
+
+
+def test_UNTLElement_add_child_exception():
+    """Test random children are not allowed."""
+    child_element = us.UNTLElement()
+    child_element.tag = 'child1'
+    element = us.UNTLElement()
+    with pytest.raises(us.UNTLStructureException):
+        element.add_child(child_element)
+
+
+def test_UNTLElement_set_content():
+    """Test setting and stripping of content value."""
+    element = us.UNTLElement()
+    element.set_content(u'  содержани ')
+    assert element.content == u'содержани'
+
+
+def test_UNTLElement_set_content_exception():
+    """Test content must be allowed to set it."""
+    element = us.UNTLElement()
+    element.allows_content = False
+    with pytest.raises(us.UNTLStructureException):
+        element.set_content('test_content')
+
+
+def test_UNTLElement_add_form_qualifier_and_content():
```

Comment:
This is how I see it looking if I use a mock:
```
@patch('pyuntl.form_logic.Title.__init__', return_value=None)
def test_UNTLElement_add_form_qualifier_and_content_mocked(mock_title):
    """Test form is created with qualifier and content passed."""
    element = us.UNTLElement()
    element.tag = 'title'
    vocabularies = {'title-qualifiers': ['test']}
    qualifier = 'test_qualifier'
    content = 'test_content'
    element.add_form(vocabularies=vocabularies,
                     qualifier=qualifier,
                     content=content)
    mock_title.assert_called_once_with(vocabularies=vocabularies,
                                       qualifier_value=qualifier,
                                       input_value=content,
                                       untl_object=element,
                                       superuser=False)
```
I am wondering though since the outcome of calling this method should be that the `form` attribute is assigned a `FormElement` object, and the above test is not checking that, is this a better test? Should I include both tests? Note: I may not be adding FormElement tests in this go around since the changes submitted for the 2->3 conversion don't show changes.

Reply:
eh, I'm not sure if it's better or not, TBH. I think this is going to come down to your own preference. You could add a second assertion that just checks the type of `element.form`, but at this point I'm really on the fence about which way comes out cleaner.

Reply:
I have now added tests that check the call as well as the correct type being set on the object.